torch/_dynamo/guards.py

```diff
@@ -349,13 +349,13 @@ def HASATTR(self, guard: Guard):
         self._produce_guard_code(guard, [code], provided_guarded_object=self.get(base))
 
     def FUNCTORCH_CURRENT_LEVEL_MATCH(self, guard: Guard):
-        # Invalidate the graph if a call to vmap has been made prior to this
-        # This is super conservative as the interpreter stack may not contain
-        # vmap
-        code = [
-            "torch._C._functorch.maybe_current_level() is None",
-        ]
-        self._produce_guard_code(guard, code)
+        # Invalidate functorch code if current level is different than
+        # the one when FX graph was generated
+        if torch._C._functorch.peek_interpreter_stack() is not None:
+            cis = torch._functorch.pyfunctorch.retrieve_all_functorch_interpreters()
+            states = [ci.get_state() for ci in cis]
+            code = [f"torch._functorch.pyfunctorch.compare_functorch_state({states})"]
+            self._produce_guard_code(guard, code)
```

Comment:
not a `FUNCTORCH_CURRENT_LEVEL_MATCH` anymore -- can we rename to `FUNCTORCH_STACK_MATCH` ?

Reply:
NB: if the stack is None, we still want to do the matching, right?

Reply:
If the stack is none, and we previously generated a graph with something on the stack, the guard should invalidate the existing graph. I'll add a test for this.